### PR TITLE
FSC Improvements

### PR DESCRIPTION
**Pull request by adecaro**
- FSC resolvers, identity.id is not needed
- FSC's tls client* keys removed

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml [hunk 1]
```diff
@@ -9,15 +9,8 @@ fsc:
...
-  # naming docker resources.
-  networkId: 5utks6sssfhvzag6vceymm4rjy
   # This represents the endpoint to other FSC nodes in the same organization.
   address: 127.0.0.1:20000
-  # Whether the FSC node should programmatically determine its address
-  # This case is useful for docker containers.
-  # When set to true, will override FSC address.
-  addressAutoDetect: true
   # GRPC Server listener address
   listenAddress: 127.0.0.1:20000
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml:15]: listenAddress could be removed now

File: integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml [hunk 2]
```diff
@@ -49,22 +42,13 @@ fsc:
...
-    # If not set, fsc.tls.key.file will be used instead
-    clientKey:
-      file: ./../../crypto/peerOrganizations/fsc.example.com/peers/initiator.fsc.example.com/tls/server.key
-    # rootcert.file represents the trusted root certificate chain used for verifying certificates
-    # of other nodes during outbound connections.
     rootcert:
       file: ./../../crypto/peerOrganizations/fsc.example.com/peers/initiator.fsc.example.com/tls/ca.crt
     # If mutual TLS is enabled, clientRootCAs.files contains a list of additional root certificates
     # used for verifying certificates of client connections.
     clientRootCAs:
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml:49]: clientRootCAs not needed as not using mutualTLS

File: integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml [hunk 3]
```diff
@@ -49,22 +42,13 @@ fsc:
...
     # If not set, fsc.tls.cert.file will be used instead
-    clientCert:
-      file: ./../../crypto/peerOrganizations/fsc.example.com/peers/initiator.fsc.example.com/tls/server.crt
-    # Private key used for TLS when making client connections.
-    # If not set, fsc.tls.key.file will be used instead
-    clientKey:
-      file: ./../../crypto/peerOrganizations/fsc.example.com/peers/initiator.fsc.example.com/tls/server.key
-    # rootcert.file represents the trusted root certificate chain used for verifying certificates
-    # of other nodes during outbound connections.
     rootcert:
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/initiator/core.yaml:45]: rootcert can be removed

File: integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml [hunk 4]
```diff
@@ -9,15 +9,8 @@ fsc:
...
-  # naming docker resources.
-  networkId: 5utks6sssfhvzag6vceymm4rjy
   # This represents the endpoint to other FSC nodes in the same organization.
   address: 127.0.0.1:20002
-  # Whether the FSC node should programmatically determine its address
-  # This case is useful for docker containers.
-  # When set to true, will override FSC address.
-  addressAutoDetect: true
   # GRPC Server listener address
   listenAddress: 127.0.0.1:20002
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml:15]: listenAddress can be removed

File: integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml [hunk 5]
```diff
@@ -47,14 +40,6 @@ fsc:
...
-    # If not set, fsc.tls.cert.file will be used instead
-    clientCert:
-      file: ./../../crypto/peerOrganizations/fsc.example.com/peers/responder.fsc.example.com/tls/server.crt
-    # Private key used for TLS when making client connections.
-    # If not set, fsc.tls.key.file will be used instead
-    clientKey:
-      file: ./../../crypto/peerOrganizations/fsc.example.com/peers/responder.fsc.example.com/tls/server.key
     # rootcert.file represents the trusted root certificate chain used for verifying certificates
     # of other nodes during outbound connections.
     rootcert:
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml:45]: rootcert can be removed

File: integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml [hunk 6]
```diff
@@ -64,7 +49,6 @@ fsc:
     clientRootCAs:
```
Comment (davidkel) [inline integration/fsc/pingpong/testdata/fsc/nodes/responder/core.yaml:49]: clientRootCAs can be removed as mutual TLS is not enabled

File: integration/nwo/client/web.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+// NewWebClientConfigFromFSC returns a web configuration from an FSC node configuration file
+func NewWebClientConfigFromFSC(confDir string) (*web.Config, error) {
+	config := &web.Config{}
+	configProvider, err := config2.NewProvider(confDir)
+	if err != nil {
+		return nil, err
+	}
+	if configProvider.GetBool("fsc.web.tls.enabled") {
+		config.URL = fmt.Sprintf("https://%s", configProvider.GetString("fsc.web.address"))
+		config.CACert = configProvider.TranslatePath(configProvider.GetStringSlice("fsc.web.tls.clientRootCAs.files")[0])
```
Comment (davidkel) [inline integration/nwo/client/web.go:25]: This line will panic if the entry is not in the core file

File: integration/nwo/fabric/topology/core_template.go [hunk 8]
```diff
@@ -277,20 +277,10 @@ fabric:
...
-        file: {{ .PeerLocalTLSDir Peer }}/server.crt
-      key:
-        file: {{ .PeerLocalTLSDir Peer }}/server.key
       clientCert:
         file: {{ .PeerLocalTLSDir Peer }}/server.crt
       clientKey:
         file: {{ .PeerLocalTLSDir Peer }}/server.key
-      rootcert:
-        file: {{ .PeerLocalTLSDir Peer }}/ca.crt
-      clientRootCAs:
```
Comment (davidkel) [inline integration/nwo/fabric/topology/core_template.go:290]: shouldn't clientRootCAs be populated if clientAuthRequired is true ?

File: integration/nwo/fsc/node/core_template.go [hunk 9]
```diff
@@ -57,24 +48,11 @@ fsc:
     # Private key used for TLS server
     key:
       file: {{ .NodeLocalTLSDir Peer }}/server.key
-    # X.509 certificate used for TLS when making client connections.
-    # If not set, fsc.tls.cert.file will be used instead
-    clientCert:
```
Comment (davidkel) [inline integration/nwo/fsc/node/core_template.go:62]: should keep clientCert and clientKey if clientAuthRequired is set to true

File: platform/fabric/core/generic/chaincode/chaincode.go [hunk 10]
```diff
@@ -69,3 +70,43 @@ func (c *Chaincode) IsPrivate() bool {
 	// Nothing was found
 	return false
 }
+
+// Version returns the version of this chaincode.
+// It uses discovery to extract this information from the endorsers
+func (c *Chaincode) Version() (string, error) {
+	response, err := NewDiscovery(c).Response()
+	if err != nil {
+		return "", errors.WithMessage(err, "failed to discover")
```
Comment (davidkel) [inline platform/fabric/core/generic/chaincode/chaincode.go:79]: I think this error needs more detail here such as failed to discover what ? what was it trying to do

File: platform/fabric/core/generic/chaincode/discovery.go [hunk 11]
```diff
@@ -127,15 +102,51 @@ func (d *Discovery) Call() ([]driver.DiscoveredPeer, error) {
 		})
 	}
 
-	// cache response
+	return discoveredEndorsers, nil
+}
+
+func (d *Discovery) Response() (discovery.Response, error) {
```
Comment (davidkel) [inline platform/fabric/core/generic/chaincode/discovery.go:108]: Not sure this method describes what is actually going on here, maybe needs to be a bit more descriptive

File: platform/fabric/core/generic/msp/x509/loader.go [hunk 12]
```diff
@@ -10,10 +10,11 @@ import (
 	x5092 "crypto/x509"
 	"encoding/hex"
 	"encoding/pem"
+	"io/ioutil"
```
Comment (davidkel) [inline platform/fabric/core/generic/msp/x509/loader.go:13]: I believe ioutil is deprecated in the Go runtime now so should avoid it

File: platform/fabric/core/generic/msp/x509/loader.go [hunk 13]
```diff
@@ -172,3 +219,34 @@ func GetEnrollmentID(id []byte) (string, error) {
...
 }
+
+func getPemMaterialFromDir(dir string) ([][]byte, error) {
+	_, err := os.Stat(dir)
+	if os.IsNotExist(err) {
+		return nil, err
+	}
+
+	content := make([][]byte, 0)
+	files, err := ioutil.ReadDir(dir)
```
Comment (davidkel) [inline platform/fabric/core/generic/msp/x509/loader.go:230]: see comment about ioutil above

File: platform/fabric/services/state/namespace.go [hunk 14]
```diff
@@ -95,7 +93,7 @@ func (n *Namespace) Present() bool {
 
 // SetNamespace sets the name of this namespace
 func (n *Namespace) SetNamespace(ns string) {
-	n.tx.SetProposal(ns, "Version-0.0", "_state")
+	n.tx.SetProposal(ns, "", "_state")
```
Comment (davidkel) [inline platform/fabric/services/state/namespace.go:96]: is a blank string still correct here ? does it get populated later ?

File: platform/view/sdk/sdk.go [hunk 15]
```diff
@@ -216,7 +212,7 @@ func (p *p) registerViewServiceServer() error {
 func (p *p) initGRPCServer() error {
 	configProvider := view.GetConfigService(p.registry)
 
-	listenAddr := configProvider.GetString("fsc.listenAddress")
+	listenAddr := configProvider.GetString("fsc.address")
```
Comment (davidkel) [inline platform/view/sdk/sdk.go:215]: Probably something that needs testing for the whole core file but what happens if fsc.address is not specified ?

File: platform/view/services/client/web/client.go [hunk 16]
```diff
@@ -39,24 +43,28 @@ type Client struct {
...
-	}
-	clientCert, err := tls.LoadX509KeyPair(
-		config.TLSCert,
-		config.TLSKey,
-	)
-	if err != nil {
-		return nil, err
+	if len(config.CACert) != 0 {
+		clientCertPool := x509.NewCertPool()
+		caCert, err := ioutil.ReadFile(config.CACert)
```
Comment (davidkel) [inline platform/view/services/client/web/client.go:50]: ioutil again

**Review by davidkel — COMMENTED** [review]
I've made a few minor comments

**Comment by adecaro** [comment]
LGTM, I'll merge as soon as the CI passes

**Pull request merged**